src/read.c

```diff
@@ -566,7 +566,7 @@ static avifResult avifCodecDecodeInputFillFromDecoderItem(avifCodecDecodeInput *
             }
         }
         if (remainingSize > 0) {
-            assert(layerCount == 3);
+            AVIF_ASSERT(layerCount == 3, AVIF_RESULT_UNKNOWN_ERROR);
```

Comment:
We should only use `assert()` for conditions that must be true by logical reasoning. We should not use `assert()` to detect bad inputs or environment errors (e.g., filesystem errors). Under this policy, an assertion failure indicates a bug in our code. If we want to return an error in that case, let's add a new error code `AVIF_RESULT_INTERNAL_ERROR`.

Another issue is that some assertion failures imply that memory may have been corrupted or our data structures may be in an inconsistent state, so it may be unsafe to return an error and continue execution. This is why in Google's internal repository and in Chrome, there is a `CHECK` that is always enabled. We may want to consider adding our equivalent of `CHECK`. It would be simpler than analyzing whether it is safe to return an error and continue execution.

Reply:
More on my suggestion of adding `AVIF_RESULT_INTERNAL_ERROR`:

`AVIF_RESULT_UNKNOWN_ERROR` is equivalent to `absl::StatusCode::kUnknown`, defined as follows:

```
  // kUnknown [...] indicates an unknown error occurred. In
  // general, more specific errors should be raised, if possible. Errors raised
  // by APIs that do not return enough error information may be converted to
  // this error.
```

`AVIF_RESULT_INTERNAL_ERROR` is equivalent to `absl::StatusCode::kInternal`, defined as follows:

```
  // kInternal [...] indicates an internal error has occurred
  // and some invariants expected by the underlying system have not been
  // satisfied. This error code is reserved for serious errors.
```

Reply:
> We should only use `assert()` for conditions that must be true by logical reasoning. We should not use `assert()` to detect bad inputs or environment errors (e.g., filesystem errors). Under this policy, an assertion failure indicates a bug in our code. If we want to return an error in that case, let's add a new error code `AVIF_RESULT_INTERNAL_ERROR`.

Done with `AVIF_CHECKERR(..., AVIF_RESULT_INTERNAL_ERROR)`.  
Thank you for checking abseil status code meanings.

> Another issue is that some assertion failures imply that memory may have been corrupted or our data structures may be in an inconsistent state, so it may be unsafe to return an error and continue execution. This is why in Google's internal repository and in Chrome, there is a `CHECK` that is always enabled. We may want to consider adding our equivalent of `CHECK`. It would be simpler than analyzing whether it is safe to return an error and continue execution.

Terminating execution may be safer but crashes are inconvenient for users.

In the example mentioned in the PR description, I believe libavif refusing to decode a few rare images (color grid but no alpha grid) is better than crashing the AVIF decoding, potentially also the whole service using libavif as a dependency. Most of the checks enforced in this PR are preventive ("alpha should be defined before being used" etc). Returning an error there seems fine.

I suggest replacing most assertions in libavif by `AVIF_CHECKERR(..., AVIF_RESULT_INTERNAL_ERROR)` and having some kind of `CHECK` in critical places, if necessary.

Reply:
Note: I just read the message as the beginning of this pull request. Your `AVIF_RESULT_ASSERTION_FAILED` is essentially  the `AVIF_RESULT_INTERNAL_ERROR` I suggested.